File: leet_code/trees/valid_binary_search_tree.js
const isValidBST = (root) => {
  if (!root || (!root.left && !root.right)) return true;
  let result = true;
  const inorder = [],
        traverse = node => {
          if (node.left) traverse(node.left);
          inorder.push(node.val);
          const last = inorder.length - 1;
          if (inorder.length > 1 && inorder[last] <= inorder[last-1]) {
            result = false;
          }
          if (node.right) traverse(node.right);
        };
  traverse(root);
  return result;
};

File: leet_code/trees/valid_binary_search_tree.rb
def is_valid_bst(root)
  return true if !root || (!root.left && !root.right)
  result = true
  inorder = []
  traverse = lambda do |node|
    traverse.call(node.left) if node.left
    inorder << node.val
    if inorder.length > 1 && inorder[-1] <= inorder[-2]
      result = false;
    end
    traverse.call(node.right) if node.right
  end
  traverse.call(root)
  result
end
